locked-asset/simple-lock-energy/src/unlock_with_penalty.rs

```diff
@@ -179,29 +193,92 @@ pub trait UnlockWithPenaltyModule:
                         && epochs_to_reduce <= lock_epochs_remaining,
                     "Invalid epochs to reduce"
                 );
-
                 epochs_to_reduce
             }
-            None => lock_epochs_remaining,
+            None => 0u64,
+        }
+    }
+
+    fn calculate_penalty_percentage_partial_unlock(
+        &self,
+        target_epoch_to_reduce: Epoch,
+        current_unlock_epoch: Epoch,
+        penalty_percentage_struct: PenaltyPercentage,
+    ) -> u64 {
+        let penalty_percentage_full_unlock_current_epoch = self
+            .calculate_penalty_percentage_full_unlock(
+                current_unlock_epoch,
+                &penalty_percentage_struct,
+            );
+        let penalty_percentage_full_unlock_target_epoch = self
+            .calculate_penalty_percentage_full_unlock(
+                target_epoch_to_reduce,
+                &penalty_percentage_struct,
+            );
+
+        (penalty_percentage_full_unlock_current_epoch - penalty_percentage_full_unlock_target_epoch)
+            * MAX_PERCENTAGE as u64
+            / (MAX_PERCENTAGE as u64 - penalty_percentage_full_unlock_target_epoch)
+    }
+
+    fn calculate_penalty_percentage_full_unlock(
+        &self,
+        current_unlock_epoch: Epoch,
+        penalty_percentage_struct: &PenaltyPercentage,
+    ) -> u64 {
+        let first_threshold_penalty = penalty_percentage_struct.first_threshold as u64;
+        let second_threshold_penalty = penalty_percentage_struct.second_threshold as u64;
+        let third_threshold_penalty = penalty_percentage_struct.third_threshold as u64;
+
+        match current_unlock_epoch / (EPOCHS_PER_YEAR + 1u64) {
+            0 => first_threshold_penalty * current_unlock_epoch / EPOCHS_PER_YEAR,
+            1 => {
+                // value between 0 and 360
+                let normalized_current_epoch_unlock = current_unlock_epoch - EPOCHS_PER_YEAR;
+                first_threshold_penalty
+                    + (second_threshold_penalty - first_threshold_penalty)
+                        * normalized_current_epoch_unlock
+                        / EPOCHS_PER_YEAR
+            }
+            2 | 3 => {
```

Comment:
2 | 3 | 4

Reply:
we can have up to 4 years.

Reply:
The calculation is `lock_epochs_remaining / (EPOCHS_PER_YEAR + 1u64)`  which means that if the token is locked for 4 years, 1440 epochs / 361 is 3.988 which is 3.
I did this so that `lock_epochs_remaining = 360` will be part of the 1 year fee, `lock_epochs_remaining=720` epochs will be part of 2 year fee and so on.